app/src/ingestion/pdf_postprocess.py

```diff
@@ -1,18 +1,67 @@
 import logging
 
 from src.ingestion.pdf_elements import EnrichedText, TextType
+from src.ingestion.pdf_stylings import Styling
+from src.util.string_utils import basic_ascii
 
 logger = logging.getLogger(__name__)
 
 
+def associate_stylings(
+    enriched_texts: list[EnrichedText], stylings: list[Styling]
+) -> list[EnrichedText]:
+    "Given EnrichedTexts and Stylings, assocate stylings to the corresponding text item"
+    for e_text in enriched_texts:
+        matched_stylings = [
+            styling for styling in stylings if styling_matches_text(styling, e_text)
+        ]
+        if matched_stylings:
+            e_text.stylings = matched_stylings
+    return enriched_texts
+
+
+def styling_matches_text(styling: Styling, e_text: EnrichedText) -> bool:
+    # Quick checks
+    if styling.pageno != e_text.page_number or styling.headings != e_text.headings:
+        return False
+
+    # Slower checks
+    stripped_wider_text = basic_ascii(styling.wider_text).strip()
+    stripped_e_text = basic_ascii(e_text.text).strip()
+    return (
+        stripped_wider_text in stripped_e_text
+        and abs(len(stripped_wider_text) - len(stripped_e_text)) < 10
+    )
+
+
+def apply_stylings(e_text: EnrichedText) -> EnrichedText:
+    "Given EnrichedTexts with stylings field, apply stylings to the text in markdown format"
+    if e_text.stylings:
+        applied = []
+        for styling in e_text.stylings:
+            markdown_text = e_text.text.replace(styling.text, f"**{styling.text}**")
+            if e_text.text != markdown_text:
+                applied.append(styling)
+                e_text.text = markdown_text
+
+        if applied == e_text.stylings:
+            e_text.stylings = None
+        else:
+            e_text.stylings = [s for s in e_text.stylings if s not in applied]
```

Comment:
Your suggestion is what I had originally but I added the `= None` line to keep the unit tests more concise and less cluttered since the unit tests compares against the original `enriched_texts`, which have `stylings = None` which is different from `stylings = []`.